libyul/backends/wasm/polyfill/Interface.yul

```diff
@@ -64,11 +65,23 @@ function calldatasize() -> z1, z2, z3, z4 {
 }
 
 function calldatacopy(x1, x2, x3, x4, y1, y2, y3, y4, z1, z2, z3, z4) {
-	eth.callDataCopy(
-		to_internal_i32ptr(x1, x2, x3, x4),
-		u256_to_i32(y1, y2, y3, y4),
-		u256_to_i32(z1, z2, z3, z4)
-	)
+	let cds:i32 := eth.getCallDataSize()
+	let offset:i32 := u256_to_i32(y1, y2, y3, y4)
+	let size:i32 := u256_to_i32(z1, z2, z3, z4)
+	// overflowed?
+	if i32.gt_u(offset, i32.sub(0xffffffff:i32, size)) {
+		eth.revert(0:i32, 0:i32)
+	}
+	if i32.gt_u(i32.add(size, offset), cds) {
```

Comment:
We should write zeros to memory for everything after cds - this detail of EVM's calldatacopy is actually used in solidity.

Reply:
added `memset 0x00`

Reply:
Is that a builtin?

Reply:
Also, memset is actually only needed inside this if statement, isn't it?

Reply:
This raises a question. The polyfill has a ton of helper functions which are not part of the dialect. Perhaps we should run a renaming step prior to the polyfill to avoid clashes.

Reply:
We already do: https://github.com/ethereum/solidity/blob/develop/libyul/backends/wasm/EVMToEwasmTranslator.cpp#L79